Create kops aws job using kubetest directly

justinsb opened the pull request:
Currently testing an otherwise unused configuration (networking=kopeio-vxlan);
also starting by just testing a small subset of tests.

krzyzacy commented:
I think you might want to call through `kubernetes_e2e.py` now, which I dumped all kops flags there, otherwise you will not have the right credentials.

justinsb commented:
Having this test actually pass is likely also blocked on https://github.com/kubernetes/test-infra/issues/3870 but it "passes on my machine" :-)

krzyzacy commented:
And will you be interested in moving logic from kops-e2e-runner.sh into kubetest/kops.go? :-)

/lint

justinsb commented on kubetest/kops.go:
```diff
@@ -31,7 +33,8 @@ import (
 	// kops specific flags.
-	kopsPath         = flag.String("kops", "", "(kops only) Path to the kops binary. Must be set for kops.")
+	kopsPath         = flag.String("kops", "", "(kops only) Path to the kops binary. kops will be downloaded from kops-base-url if not set.")
+	kopsBaseUrl      = flag.String("kops-base-url", os.Getenv("KOPS_BASE_URL"), "(kops only) binary distribution of kops to run.")
```
Golint naming: Should you be called GoLint?

justinsb commented:
/lint

justinsb commented:
@krzyzacy yes - so I got fejta to explain the philosophy to me, and my understanding is that _if_ we can get it into kubetest/kops.go, and simply pass-through flags through execute.py, that is best.  But we shouldn't be building in kubetest, and it's not clear that we can through execute.py.

But yes, particularly here, I've tried putting some of the logic that is currently in kops-e2e-runner.sh into kubetest/kops.go.

I'm hoping that we can have prow mount the AWS credentials file, and pass an env var to the location (if needed), just as we do in that excerpt I pulled out for GOOGLE_APPLICATION_CREDENTIALS in #3870.

If I've got the wrong end of the stick here, let me know!

krzyzacy commented:
I'm moving towards https://github.com/kubernetes/test-infra/pull/3682, which will make all e2e jobs more consistent :-)

I'm more than happy to set up the prow secrets, but I'm not sure I can get the raw credential from Jenkins? Or where should I?

fejta commented:
My main request is that we use bootstrap.py to check out all the git repos we need and to not make git clone calls inside kubetest.

There is already a --build flag to build kubernetes, and a --kops-build or --build-kops sounds legit too -- potentially better as then the build will happen inside a sanitized environment (either inside a docker container or kubernetes container) rather than outside on a VM.

justinsb commented:
That makes sense @fejta ... at least in this scenario I am testing a version of kops and kubernetes that has already been built (the latest CI for each).  I was thinking regardless of scripts, it makes sense to build in a separate step from e2e testing.  kops doesn't use bazel at the moment, but I'm looking at that model.

So in this PR we go through bootstrap.py, then use execute.py, then jump into kubetest.  I added the ability to kubetest to download a prebuilt kops version, just as we can download a prebuilt k8s version.  I also added the ability to specify a kubernetes version to kops using `ci/latest`, which I'm less happy about because it duplicates the k8s resolve/download code, but it isn't _too_ bad I don't think.

I think I'm starting to understand the vision - that we have a test pod that can run against any version of the code.  It's very powerful even if it is taking me a while to grok it - thanks for the patience in explaining it!

krzyzacy commented on jobs/config.json:
```diff
@@ -8074,6 +8074,28 @@
+      "--test_args=--ginkgo.focus=ConfigMap --ginkgo.skip=\\[Slow\\]|\\[Serial\\]|\\[Disruptive\\]|\\[Flaky\\]|\\[Feature:.+\\]",
+      "--v"
+    ],
+    "scenario": "execute",
```
I think you'll still want to use `kubernetes_e2e` scenario here, unless we don't care anything in https://github.com/kubernetes/test-infra/blob/master/jobs/platform/kops_aws.env anymore? Looks like it will not dump logs properly.

justinsb replied:
Good point.  I would say that I didn't consider these logs, rather than I don't think it matters :-)

I am starting with a new e2e test, rather than risking one of the existing ones...

krzyzacy commented on prow/config.yaml:
```diff
@@ -13743,6 +13743,29 @@ periodics:
+  name: ci-kubernetes-e2e-kops-aws-kopeiovxlan
+  spec:
+    containers:
+    - image: gcr.io/k8s-testimages/kubekins-e2e-prow:v20170731-2e38747a
```
you'll need to commit kubetest changes first, then build a new image from there :-)

krzyzacy commented on prow/config.yaml:
```diff
@@ -13743,6 +13743,29 @@ periodics:
+    containers:
+    - image: gcr.io/k8s-testimages/kubekins-e2e-prow:v20170731-2e38747a
+      args:
+      - "--json"
```
`--json` is by default in the image, you don't need it here

justinsb commented:
Re the kubetest changes needing a new image, should I therefore split this into two PRs (kubetest & the job itself)?

krzyzacy commented:
You just need to split it into multiple commits :-)

justinsb commented:
Split into two commits!  Obviously I'm still figuring this out, but I know how to predict the image name though, so the second commit really doesn't have much hope of working (even without the known problems with the AWS key / logging etc)

krzyzacy commented:
okay, @justinsb I added two secrets `aws-cred` and `aws-ssh-key-secret` to prow, you can update prow config now and see if they will work :-)

krzyzacy commented:
ref https://github.com/kubernetes/test-infra/pull/3997

justinsb commented:
@krzyzacy thanks for uploading the keys.  I updated the prow spec to reference the new keys following your example.  Sorry for delay - was on holiday :-)

justinsb commented:
ping @krzyzacy can we get this merged so I can start seeing what breaks? :-)

krzyzacy commented:
I was playing around with aws-canary job which runs in prow now, and I seems still have some ssh key issue.

I think you want to mimic https://github.com/justinsb/test-infra/blob/bb40a97309b9dff4c19f3e7ac0e948e3b95cd179/jobs/config.json#L8248-L8263 for your job here.

krzyzacy commented:
@justinsb since you are around - when I convert kops-aws jobs to use latest everything, I got a
`ResumableUploadAbortException: 403 661044217466-compute@developer.gserviceaccount.com does not have storage.objects.create access to bucket kops-ci.`

log: https://k8s-gubernator.appspot.com/build/kubernetes-jenkins/logs/ci-kubernetes-e2e-kops-aws/15893

Can it be some configuration you can fix? Or it's from other causes?

justinsb commented:
Big picture question: is it OK to try to use kubetest directly?  As an outsider that is less familiar with the infrastructure, I found kubetest much simpler to grasp (I was able to run it standalone locally with no wrapper scripts, for example).

But it seems that you're suggesting @krzyzacy that I should be using kubernetes_e2e (?).  I do see that kubernetes_e2e offers more functionality (e.g. logs as you pointed out).

So I guess the question is whether I should continue to pursue the (IMO) more understandable "kubetest-direct" approach, or whether I should abandon this PR and try to use kubernetes_e2e?

My preference would be to get kubetest built up with any missing functionality - as long as it remains simple.

I guess cc @fejta as well, though I worry I'll get different answers :-)

krzyzacy commented:
kubernetes_e2e takes in all kubetest flags. It sets up aws ssh/credentials to the right place so you'll still need that..

basically, it will look like
```
  "ci-kubernetes-e2e-kops-aws-kopeiovxlan": {
    "args": [
      "--deployment=kops",
      "--extract=ci/latest",
      "--kops-kubernetes-version=ci/latest",
      "--kops-cluster=e2e-kops-aws-kopeiovxlan.test-aws.k8s.io",
      "--kops-args=--networking=kopeio-vxlan",
      "--kops-state=s3://k8s-kops-jenkins/",
      "--test_args=--ginkgo.focus=ConfigMap --ginkgo.skip=\\[Slow\\]|\\[Serial\\]|\\[Disruptive\\]|\\[Flaky\\]|\\[Feature:.+\\]",
    ],
    "scenario": "kubernetes_e2e.py",
    "sigOwners": [
      "sig-aws",
      "sig-cluster-lifecycle"
    ]
  },
```

krzyzacy commented:
(If you want to try this way I can get it merged and we can see what's missing)

fejta commented:
Calling kubetest directly from bootstrap.py is fine... however it will be easier for other people to maintain if we follow the standard pattern that every other job uses, meaning use the scenario

justinsb commented:
I would love to try this way if that's OK.  Anything beyond kubetest is very hard for people like me that are not familiar with sig-testing.

I would love to get to the point where we can give someone a kubetest command, and say "here's how you run the test".  Yes, there might be 40 flags on there, but that's OK because it's explicit and discoverable.  And then we give them a standalone pod manifest which they can run, or maybe even wrap in a CronJob and submit to the centralized dashboard if they want to run the tests themselves.

I do hear the arguments about the value of the scenarios to team-test, but that value to me _personally_ as an outsider is outweighed by their (seeming) complexity.

Of course, I fully expect that if we merge this, I'll grok what the scenarios are doing, and then I'll want to move to scenarios.  Not sure how to deal with that one... :-)

krzyzacy commented:
@justinsb the scenario is a nice wrapper for kubetest which it reduces the hassle for your 40 kubetest flags  and it's easier to configure. Also because all other jobs are working that way now, I'd like to keep the consistency.

We can merge for now, and I'll let you figure out the scenario side.

** Calling kubetest directly locally is fine, as long as people have right stuff in right places that will work, but as a ci system we need to set up credentials/collect artifacts, so a layer for ci is necessary. You can still hand people the kubetest command, which you can find from the build log.

krzyzacy commented:
/lint

fejta commented:
You will spend less time now and in the future if you just use the scenario like every other job.

The scenario just sets environment variables, and does a bunch of irrelevant stuff for docker which makes it look more complicated.

The reason you are in this mess in the first place is because kops isn't following the standard patterns

krzyzacy commented:
btw, once we sort out https://github.com/kubernetes/test-infra/pull/4254, all kops-aws jobs can keep in sync with rest of the e2e jobs and they will be in a better world.

krzyzacy commented:
/lgtm

@justinsb let me know if you need me to build a new image for you

justinsb commented:
@fejta I don't understand.  You want everything to go to kubetest (?), and I believe am doing that with this job.

Thanks @krzyzacy - I believe I am using `gcr.io/k8s-testimages/kubekins-e2e-prow:latest` but if that version isn't (periodically?) built from master then a new version would be greatly appreciated, thank you.

krzyzacy commented:
@justinsb if you expect to change kubetest for kops part, using :latest is fine and it's easier to push a :latest tag without affecting prod jobs. However that tag can be overwritten pretty easily.

justinsb commented:
Closing this - we have the bulk of the kubetest changes in for other tests, though using kubenetes-e2e
